### PR TITLE
fix: add activeness check in scenario runner

### LINKED ISSUE #14186: Script service is missing an activeness check for disclosed contracts
This should fail but it succeeds
```
module Main where

import Daml.Script

template T
  with
    p : Party
  where
    signatory p
    nonconsuming choice Fetch : T
      with
        p' : Party
      controller p'
      do pure this

template Helper
  with
    p : Party
  where
    signatory p
    choice Use : ()
      with
        cid : ContractId T
      controller p
      do exercise cid (Fetch p)
         pure ()


test = script do
  p1 <- allocateParty "p1"
  p2 <- allocateParty "p2"
  cid <- submit p1 $ createCmd (T p1)
  Some inst <- queryContractId' p1 cid
  let disclosed = [toDisclosedContract cid inst.contract inst.metadata]
  submitDisclosing disclosed p2 $ createAndExerciseCmd (Helper p2) (Use cid)
  submit p1 $ archiveCmd cid
  submitDisclosing disclosed p2 $ createAndExerciseCmd (Helper p2) (Use cid)
  pure ()
```

### PR DESCRIPTION
**Pull request by ghost**
This adds an activeness check for disclosed contracts.

This only becomes relevant, when the disclosure PoC branch gets merged into main.

TODO: Tests?

Fixes #14186.

CHANGELOG_BEGIN
CHANGELOG_END

### CODE REVIEW

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala [hunk 1]
```diff
@@ -436,16 +436,16 @@ sealed abstract class HasTxNodes {
...
+    fold(Set.empty[(Cid2, Identifier)]) {
+      case (acc, (_, Node.Exercise(coid, tid, _, _, _, _, _, _, _, _, _, _, _, _, _))) =>
+        acc + ((coid, tid))
+      case (acc, (_, Node.Fetch(coid, tid, _, _, _, _, _, _))) =>
+        acc + ((coid, tid))
+      case (acc, (_, Node.LookupByKey(tid, _, Some(coid), _))) =>
+        acc + ((coid, tid))
       case (acc, _) => acc
-    } -- localContracts.keySet
+    }.filterNot(x => localContracts.keySet.contains(x._1))
```
Comment (remyhaemmerle-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala:448]: This is a public API (by KV at least). You cannot change it like that.

Comment (robin-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala:448]: Is there any other way I can get the template id? It's only used for the error message, but the message is fairly meaningless if you don't know the template of the inactive contract.

Comment (remyhaemmerle-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala:448]: Not really :(. i would probably copy/adapt your new `inputContracts` method directly you wrote in the Scenario Ledger.

 ```
 def tmplId(tx: Tx, coid: ContractId): Option[TypeConName] =
      transaction.nodes.values.collectFirst {
        case Node.Create(`coid`, tid, _, _, _, _, _, _) => tid
        case Node.Exercise(`coid`, tid, _, _, _, _, _, _, _, _, _, _, _, _, _) => tid
        case Node.Fetch(`coid`, tid, _, _, _, _, _, _) => tid
        case Node.LookupByKey(tid, _, Some(`coid`), _) => tid
      }
```

Comment (robin-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala:448]: @carlpulley-da could you please take care of this?

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/scenario/ScenarioLedger.scala [hunk 2]
```diff
@@ -386,22 +393,80 @@ object ScenarioLedger {
...
         case Some(duplicateKey) =>
-          Left(UniqueKeyViolation(duplicateKey))
+          Left(CommitError.UniqueKeyViolation(UniqueKeyViolation(duplicateKey)))
 
         case None =>
           Right(())
       }
     }
 
+    def inactiveContractCheck(ledgerData: LedgerData): Either[CommitError, Unit] = {
```
Comment (carlpulley-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/scenario/ScenarioLedger.scala:412]: Potential solution that uses the transaction to identify the contracts template

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/scenario/ScenarioLedger.scala [hunk 3]
```diff
@@ -386,22 +393,80 @@ object ScenarioLedger {
...
+            val templateId = searchForTemplate(contractId).getOrElse(
+              crash(s"failed to locate template for $contractId")
+            )
+
+            CommitError.ContractNotActive(ContractNotActive(contractId, templateId))
+        }
+        .fold[Either[CommitError, Unit]](Right(()))(Left(_))
+    }
+
+    // FIXME: alternative approach to searching based on ledgerData (potentially cleaner solution?)
```
Comment (carlpulley-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/scenario/ScenarioLedger.scala:438]: Potential solution that uses the ledger to identify the contacts template

**Comment by carlpulley-da** [comment]
PR work temporarily parked

**Pull request closed without merging**